### ISSUE TITLE
rsyslog stops logging from journal when timejump backwards, drop due to rate limit

### ISSUE THREAD

**Issue by hese10** [rsyslog/rsyslog#1354]
Using fedora 25 with systemd 232 + rsyslog 8.23-0 + systemd-timesyncd service masked. normal user added to system (user name "uuno"). Using imjournal with ratelimit values:
```
$ModLoad imjournal # provides access to the systemd journal
$imjournalRatelimitInterval 1
$imjournalRatelimitBurst 20000
$imjournalPersistStateInterval 100
$imjournalStateFile /var/lib/rsyslog/imjournal.state
# setting escaping off to make it possible to remove the control characters
$EscapeControlCharactersOnReceive off

# removing the optimization from use (it slows things down)
$OptimizeForUniprocessor off

# Using queue for 20000 messages. After that the messages are dropped instantly
$MainMsgQueueSize 20000
$MainMsgQueueDiscardMark 20000
$MainMsgQueueTimeoutEnqueue 0

```
then problem comes in when making following steps. 
1) login user "uuno"
2) write to journal with logger
3) jump system time to 2 days future with command "date -s"
4) start endless loop to write to journal as user "uuno":
`count=0; while true; do  count=$[$count+1]; logger "text $count"; sleep 0.0001;  done
`
5) wait for journal files to rotate once
6) change time back to current time with command "date -s"
7) wait for journal files to rotate once (this rotate event maybe even not needed to be waited, but from time jump point some time is needed to be waited before rsyslog reports next step problem)
8) in journal it can be seen that rsyslog drops all new messages:
Jan 10 13:33:31 vm000949 liblogging-stdlog[30900]: imjournal: begin to drop messages due to rate-limiting
rsyslog stays in this drop state forever, even after endless loop started in step 4 is stopped.
Only by restarting rsyslog service, the logging resumes.

rsyslog gets confused when time jumps backwards for rate limit calculations?

journalctl -f command however does not have this problem (with or without rate limits set in journald.conf), so shuold imjournal changed to use journalctl logic somehow?

**Comment by davidelang** [comment]
see #1290 This is a bug in the journal library code that rsyslog is using

**Comment by hese10** [comment]
but how does imjournal things that the rate limit values exceeded? as in that other case there was no logs about rate limit happening. And this stuck problem does not happen with "journalctl -f" command. so imjournal is not using same routines than "journalctl -f", right ?

**Comment by hese10** [comment]
and what is the linked systemd issue, as there is no link to it in that #1290 ... is it https://github.com/systemd/systemd/issues/1752 ?

**Comment by davidelang** [comment]
I didn't link to a systemd bug report, but rather to the RHEL bugzilla report

David Lang

**Comment by hese10** [comment]
I put more variables into that drop error log in ratelimit.c withinRatelimit function:
Jan 11 12:48:03 vm000949.nsn-intra.net liblogging-stdlog[3248]: imjournal: begin to drop messages due to rate-limiting: begin=1484870488, interval=1, tt=1484131683, burst c=20000, done c=20000

and this indicate that in my case the problem is not journal API, but c-code "time" routine. as this ratelimit.c withinRatelimit function always gets time from system "time (NULL)" call, not from input parameters.

**Comment by davidelang** [comment]
Keep in mind that there are two time values that rsyslog maintains for every log 
message

1. the time the log message was created

2. the time that rsyslog first processed the message.

for local messages, they are usually the same (although if messages are delayed 
going through journald they could be different)

which of these two timestamps are you talking about?

In any case, getting these times wrong would not result in rsyslog not 
processing messages.

David Lang

**Comment by hese10** [comment]
if we add checking for  "tt < ratelimit->begin" and set the begin values to zero, this will correct my situation.
```
        /* resume if we go out of time window */
        if(tt > ratelimit->begin + ratelimit->interval) {
                ratelimit->begin = 0;
                ratelimit->done = 0;
                tellLostCnt(ratelimit);
        } else if ( tt < ratelimit->begin ) {
                ratelimit->begin = 0;
                ratelimit->done = 0;
                tellLostCnt(ratelimit);
       }

```

**Comment by hese10** [comment]
I am talking about the timestamp that is read inside withinRatelimit function.

**Comment by rgerhards** [comment]
So far I could only have a quick look at this thread. I thnink what @hese10 explains is that the rate limiter is set to n messages within a time interval. Now the time is set backwards. IMHO would **should** happen now is that indeed the rate-limiter is reset. But we need to make sure that this is based on system time, not on any time from a message property. I do not right at the moment know which timestamp is checked in this case here, but wanted to state this comment to make things easier once we move to applying a patch.

Most importantly, the suggestion from @hese10 might already be the full solution, but we need to check it actually bases on time obtained by the API from the local system via a call from rsyslog itself. I am 80% sure this is the case.

**Comment by hese10** [comment]
I added the checking for "else if ( tt < ratelimit->begin )"
```
        /* resume if we go out of time window */
        if(tt > ratelimit->begin + ratelimit->interval) {
                ratelimit->begin = 0;
                ratelimit->done = 0;
                tellLostCnt(ratelimit);
        } else if ( tt < ratelimit->begin ) {
                ratelimit->begin = 0;
                ratelimit->done = 0;
                tellLostCnt(ratelimit);
       }

```

and the fault does not occur in my case.

**Comment by rgerhards** [comment]
@hese10 I have looked a bit at the code and the proposed fix looks well. Would you be willing to create a pull request? If not, no problem, I'll do in your name.

**Comment by rgerhards** [comment]
mhh, no hold on... let me check some journal specifcs...

**Comment by hese10** [comment]
I can make the pull request. What journal specific is there? you mean that this ( tt < ratelimit->begin ) shuold be only checked when ratelimit->bNoTimeCache = 1  ? But isn't this correction always good to have?

**Comment by radosroka** [comment]
I've already got a patch that resolves problem with systemd and time jumped backwards. I hadn't time to put it into fadora stable repository but I will do it soon. This fix probably will come with 8.24.0 update for fedora.

I will make a PR ASAP.

**Comment by hese10** [comment]
is the patch same as mine or something else?

**Comment by rgerhards** [comment]
I got dragged away while reviewing. The problem was/is more that I am not sure if we have a problem with the current imjournal code which prevents us on resetting. All other sources are OK. We cannot permit an app-generated timestamp to reset the ratelimiter. All other inputs rely on rsyslog-obtained time from OS API, which we can trust. I have now checked ```sd_journal_get_realtime_usec()``` and it looks (though not written explicitely) that this timestamp is recorded by the journal itself, which I assume took it from the OS. So this is also trustworthy. In that case, we can also trust it and so we can use the logic as suggested.

As a detail, a simple OR would probably do the job.

**Comment by hese10** [comment]
but in imjournal case,  ratelimit->bNoTimeCache = 1, thus the time get from sd_journal_get_realtime_usec() is not used inside withinRatelimit function even that goes as input parameter to it. So only "time(NULL)" is used to get the time.  or is ratelimit->bNoTimeCache not 1 in imjournal case? At least I saw it set to that value in imjournal.c code.

**Comment by rgerhards** [comment]
@hese10 jupp, you are right -- post vacation dumbness on my side ;-) The comment even says that [and I misread it ;-)]

**Comment by hese10** [comment]
ok :) Can we make the change as generic in that withinRatelimit function? e.g. not check that ratelimit->bNoTimeCache = 1, but just check if that  (tt < ratelimit->begin)

**Comment by rgerhards** [comment]
> Can we make the change as generic in that withinRatelimit function? e.g. not check that ratelimit->bNoTimeCache = 1, but just check if that (tt < ratelimit->begin)

Yes, definitely. We just need to keep the existing code that queries the system time if message from imjournal arrive.

**Comment by hese10** [comment]
Ok. But I will wait for  @radosroka to respond so that it is confirmed if same correction is already coming or not.

**Comment by hese10** [comment]
> I've already got a patch that resolves problem with systemd and time jumped backwards. I hadn't time to put it into fadora stable repository but I will do it soon. This fix probably will come with 8.24.0 update for fedora.

actually what is this now referring to? In systemd 232 there was time jump backwards bug already corrected, but that does not help totally with rsyslog. As we now we have this rsyslog imjournal code side bug that is not related at all to systemd.
